File: 0001-two-sum/0001-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    
    const hashtable = {}
    const program = new Program(nums, target, hashtable);
    const result = program.twoSum();
    
    return result;
};

class Program {
    
    constructor(nums, target, hashtable) {
        this.nums = nums;
        this.target = target;
        this.hashtable = hashtable;
    }
    
    inHashtable(num) {
        return this.hashtable[num] != null;
    }
    
    storeInverseAndIndexInHashtable(key, value) {
        this.hashtable[key] = value;
    }
    
    twoSum() {
        
        for (let i = 0; i<this.nums.length; i++) {
            
            const inverse = this.target - this.nums[i];
            
            // check if nums[i] was an inverse stored in hashtable and return answer,
            // else store inverse in hashtable.
            if (this.inHashtable(this.nums[i])) {
                return [this.hashtable[this.nums[i]], i];
            } else {
                this.storeInverseAndIndexInHashtable(inverse, i);
            }
        }
    }
    
}

File: 0001-two-sum/0001-two-sum.cs
public class Solution 
{
    public int[] TwoSum(int[] nums, int target) 
    {
        Dictionary<int, int> table = new Dictionary<int, int>();
        int[] results = new int[2];
        
        for (int i = 0; i<nums.Length; i++)
        {
            int difference = target - nums[i];
            
            if (table.ContainsKey(difference))
            {
                results[0] = (int)table[difference];
                results[1] = i;
                
                return results;
            }
            
            if (table.ContainsKey(nums[i]))
                continue;
            
            table.Add(nums[i], i);
        }
        
        return results;
    }
}